content/blog/2022-05-18-learned-federated-query-optimization.md

```diff
@@ -0,0 +1,178 @@
+AND B.id = C.id
+```
+
+If we think of this query as a graph, it would look as follows:
+```
+    A - B
+    \   / 
+      C
+```
+
+This is a fully-connected graph, as there are connections from all nodes to all nodes. As the final plan will be a binary tree,
+that means that the final plan could be any of the following three trees:
+```
+1.  ⋈         2.  ⋈          3.  ⋈ 
+   /  \          /  \           /  \
+  A   ⋈         B   ⋈          C   ⋈ 
+     / \           /  \           /  \
+    B   C         A    C         A    B
+```
+
+However, we are mostly interested in plans that look like 1. The reason is that we are mainly interested in joins that 
+can co-locate tables thus, so execution push-down is possible. Thus, we created a graph structure in which each node
+represents a table, and each edge represents a join between two tables. We enrich more our graph representation by adding
+information about the location of each table in the node, that is, in which data source the node resides. This allows
+us to develop an algorithm can can parse this graph and generate possible trees that join tables that reside in the same
+engine, when that's possible given the input query semantics (joins).
+
+### Final Plan Generation
+After processing the graph, we get a tree that collocates tables residing in the same engine by producing a tree-plan
+that joins together the collocated tables. Then, we use our machine learning model in order to predict the execution cost
+of joining the collocated tables in the machine that they reside on in the federation engine. We then produce the final
+query execution plan which also contains information about where each node of the tree will be executed.
+
+# What were the hardest parts to get right?
+The trickiest part of the project, which is still not completely resolved is the cost estimation. The average accuracy of 
+our model in predicting the cost of a given query is around 60%. However, in most of the cases in which we mispredict
+the cost, the final plan that splits end distribues the query execution across the different engines is still better
+than just using the federation engine for the query execution. We present some results that demonstrate this in the 
+final section.
+
+# Evaluation
+In order to evaluate our system we use two benchmarks, [TPC-H](https://www.tpc.org/tpch/) and 
+[Join-Order-Benchmark](https://github.com/gregrahn/join-order-benchmark). The experimental setup consisted of
+one federation engine (Spark SQL) and two external systems, MySQL and PostgreSQL. 
+
+## Cost Model Accuracy
+We first try to evaluate the accuracy
+of our machine learning cost model. For this experiment, we evaluate how well the cost model does in selecting the
+appropriate execution engine (Spark SQL, Postgres or MySQL) in order to execute some part of the query. As mentioned in 
+the previous section, the cost model accuracy is around 60%. However, in this experiment our model mispredicts in cases
```

Comment:
I feel like it might be important to know the top-1 accuracy too: that is, not just how close the prediction is to the correct cost, but how often the predictor agrees with the ground-truth about the minimum-cost plan.